### PR TITLE
Task3

### PR DESCRIPTION
**Pull request by MorozVladislav**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -21,12 +23,19 @@ def wrapped(self, *args, **kwargs):
...
                 self.coordinates[point[0]] = (x, y)
         return func(self, *args, **kwargs)
+
     return wrapped
 
 
 class Application(Frame, object):
     """The application main class."""
 
+    OBJ_TYPE = AttrDict({
```
Comment (MorozVladislav) [inline app.py]: I suppose the dict could be called simply TYPES. This may help to make some lines shorter and more readable

File: app.py [hunk 2]
```diff
@@ -48,7 +57,8 @@ def __init__(self, master=None):
         self.x0, self.y0, self.scale_x, self.scale_y = self.WIDTH / 2, self.HEIGHT / 2, None, None
         self.r = int(0.05 * min(self.x0, self.y0))
         self.font_size = self.r / 2
-        self.coordinates, self.weights, self.ids = {}, [], []
+        self.coordinates = {}
+        self.canvas_obj = {self.OBJ_TYPE.POINT:{},self.OBJ_TYPE.LINE:{},self.OBJ_TYPE.WEIGHT:{}}
```
Comment (MorozVladislav) [inline app.py]: There is no need to create self.canvas_obj with some empty internal structure. I suppose you can easily create this structure dynamically by assigning some value to necessary key. Like this:
self.canvas_obj[self.OBJ_TYPE.POINT] = <some_value>

File: app.py [hunk 3]
```diff
@@ -98,7 +109,8 @@ def graph(self, value):
 
         self.canvas.delete('all')
         self.scale_x, self.scale_y = None, None
-        self.coordinates, self.weights, self.ids = {}, [], []
+        self.coordinates = {}
+        self.canvas_obj = {self.OBJ_TYPE.POINT: {}, self.OBJ_TYPE.LINE: {}, self.OBJ_TYPE.WEIGHT: {}}
```
Comment (MorozVladislav) [inline app.py]: The same thing here. You can create self.canvas_obj like this:
self.canvas_obj = {}

File: app.py [hunk 4]
```diff
@@ -135,8 +147,9 @@ def create_points(self, points):
 
         for point in points:
             x, y = self.coordinates[point[0]]
-            self.canvas.create_oval(x - self.r, y - self.r, x + self.r, y + self.r, fill=self.POINT_COLOR)
-            self.canvas.create_text(x, y, text=point[0], font="{} {}".format(self.FONT, self.font_size))
+            point_id = self.canvas.create_oval(x - self.r, y - self.r, x + self.r, y + self.r, fill=self.POINT_COLOR)
+            text_id = self.canvas.create_text(x, y, text=point[0], font="{} {}".format(self.FONT, self.font_size))
+            self.canvas_obj[self.OBJ_TYPE.POINT][point_id] = { 'idx': point[0], 'text_id': text_id}
```
Comment (MorozVladislav) [inline app.py]: Revise this line taking into account self.canvas_obj initialization. And also consider code style.
{ 'idx': point[0]... - not cool;
{'idx': point[0]... - better

File: app.py (hunk 2 above)
Comment (MorozVladislav) [reply app.py]: Also I suppose we don't need a separate self.OBJ_TYPE.WEIGHT:{} dict. We can store weight as line attribute 'weight_obj' for example containing a tuple of ids (oval_id, text_id). What do you think of such approach?

File: app.py [hunk 5]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
         if len(self.weights) == 0:
             return
         if self.show_weight.get():
-            for weight in self.weights:
-                x, y, value = weight
+            for line in self.canvas_obj[self.OBJ_TYPE.LINE].values():
+                x_start,y_start = self.coordinates[line['point_start']]
```
Comment (MorozVladislav) [inline app.py]: Consider code style:
x_start,y_start - not cool;
x_start, y_start - better

File: app.py [hunk 6]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                x_end, y_end = self.coordinates[line['point_end']]
+                x, y = self.medium_coordinates(x_start,y_start,x_end, y_end)
```
Comment (MorozVladislav) [inline app.py]: The same thing. Follow PEP8 code style.
(x_start,y_start,x_end, y_end) - not cool;
(x_start, y_start, x_end, y_end) - better

File: app.py [hunk 7]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                value = line['weight']
                 r = int(self.r / 2) * len(str(value))
-                self.ids.append(self.canvas.create_oval(x - r, y - r, x + r, y + r, fill=self.BG, width=0))
-                self.ids.append(self.canvas.create_text(x, y, text=value, font="{} {}".format(self.FONT, str(r))))
+                weight_id = self.canvas.create_oval(x - r, y - r, x + r, y + r, fill=self.BG, width=0)
+                text_id = self.canvas.create_text(x, y, text=value, font="{} {}".format(self.FONT, str(r)))
+                self.canvas_obj[self.OBJ_TYPE.WEIGHT][weight_id] = {'text_id':text_id}
```
Comment (MorozVladislav) [inline app.py]: I suppose you can store oval_id and text_id of some weight as a line attribute. See comment above. And also code style here:
{'text_id':text_id} - not cool;
{'text_id': text_id} - better

File: app.py [hunk 8]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                line['weight_id'] = weight_id
         else:
-            for element_id in self.ids:
-                self.canvas.delete(element_id)
+            for key, values in self.canvas_obj[self.OBJ_TYPE.WEIGHT].iteritems():
```
Comment (MorozVladislav) [inline app.py]: Why do you need .iteritems()? I suppose you can use .items() instead. Also variables are usually called key and value

File: app.py [hunk 9]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                self.canvas.delete(key)
+                self.canvas.delete(values['text_id'])
+
+    def medium_coordinates(self,x_start,y_start,x_stop,y_stop):
```
Comment (MorozVladislav) [inline app.py]: For future similar lines I will just comment PEP8. That means: check code style here.

File: app.py [hunk 10]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+        return (x_start + x_stop) / 2, (y_start + y_stop) / 2
+
+    def movePoint(self, event):
```
Comment (MorozVladislav) [inline app.py]: Usually methods are called without using camel case. Rename it to move_point.

File: app.py [hunk 11]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+        x = self.canvas.canvasx(event.x)
+        y = self.canvas.canvasy(event.y)
+        obj_id = self.canvas.find_overlapping(x - 5, y - 5, x + 5, y + 5)
+        for obj in obj_id:
+            if obj in self.canvas_obj[self.OBJ_TYPE.POINT].keys():
+                obj_number = self.canvas_obj[self.OBJ_TYPE.POINT][obj]['idx']
+                self.canvas.coords(obj, event.x - self.r, event.y - self.r, event.x + self.r, event.y + self.r)
+                self.canvas.coords((self.objDict[obj][0]), event.x, event.y)
+                
+                for key, values in self.canvas_obj[self.OBJ_TYPE.LINE].iteritems():
```
Comment (MorozVladislav) [inline app.py]: The same thing: key, value and .items()

File: app.py [hunk 12]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                    new_point_x = 0
+                    new_point_y = 0
```
Comment (MorozVladislav) [inline app.py]: You can make it like this:
new_point_x, new_point_y = 0, 0
Shorter and nicer in my point of view)

File: app.py [hunk 13]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                    # (11, 5, 113.0, 216.0, 61.0, 402.0) -> values
```
Comment (MorozVladislav) [inline app.py]: Nobody will ever understand the meaning of this comment) Remove it.

File: app.py [hunk 14]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                    if obj_number in (values['point_start'],values['point_end']):
```
Comment (MorozVladislav) [inline app.py]: PEP8

File: app.py [hunk 15]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                self.canvas.delete(values['text_id'])
+
+    def medium_coordinates(self,x_start,y_start,x_stop,y_stop):
+        return (x_start + x_stop) / 2, (y_start + y_stop) / 2
+
+    def movePoint(self, event):
+        # move the point and edges
+
+        x = self.canvas.canvasx(event.x)
+        y = self.canvas.canvasy(event.y)
```
Comment (MorozVladislav) [inline app.py:212]: x, y = self.canvas.canvasx(event.x), self.canvas.canvasy(event.y)

File: app.py [hunk 16]
```diff
@@ -152,8 +165,10 @@ def create_lines(self, lines):
         for line in lines:
             x_start, y_start = self.coordinates[line[0]]
             x_stop, y_stop = self.coordinates[line[1]]
-            self.canvas.create_line(x_start, y_start, x_stop, y_stop)
-            self.weights.append(((x_start + x_stop) / 2, (y_start + y_stop) / 2, line[2]['weight']))
+            line_id = self.canvas.create_line(x_start, y_start, x_stop, y_stop)
+            self.canvas_obj[self.OBJ_TYPE.LINE][line_id] =  {'idx': line[2]['idx'], 'weight': line[2]['weight'],
```
Comment (MorozVladislav) [inline app.py]: Remove double space after =

File: app.py [hunk 17]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+        # move the point and edges
+
+        x = self.canvas.canvasx(event.x)
+        y = self.canvas.canvasy(event.y)
+        obj_id = self.canvas.find_overlapping(x - 5, y - 5, x + 5, y + 5)
+        for obj in obj_id:
+            if obj in self.canvas_obj[self.OBJ_TYPE.POINT].keys():
+                obj_number = self.canvas_obj[self.OBJ_TYPE.POINT][obj]['idx']
+                self.canvas.coords(obj, event.x - self.r, event.y - self.r, event.x + self.r, event.y + self.r)
+                self.canvas.coords((self.objDict[obj][0]), event.x, event.y)
```
Comment (MorozVladislav) [inline app.py]: How can you access self.objDict if you even don't have it? Did you try to run the code before commit? This line seems to throw error every time. Do you need it?

File: app.py (hunk 14 above)
Comment (MorozVladislav) [reply app.py]: I suppose it is better to use OR here. Like this:
if obj_number in values['point_start'] or obj_number in values['point_end']:

File: app.py [hunk 18]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                for key, values in self.canvas_obj[self.OBJ_TYPE.LINE].iteritems():
+                    new_point_x = 0
+                    new_point_y = 0
+                    # (11, 5, 113.0, 216.0, 61.0, 402.0) -> values
+                    if obj_number in (values['point_start'],values['point_end']):
+                        if obj_number == values['point_start']:
+                            new_point_x, new_point_y = self.coordinates[values['point_end']]
+                            self.canvas.coords(key, x, y, new_point_x, new_point_y)
+
+                        elif obj_number == values['point_end']:
```
Comment (MorozVladislav) [inline app.py]: Why do you need elif here? May be simple else is enough?

File: app.py [hunk 19]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                            new_point_x, new_point_y = self.coordinates[values['point_start']]
+                            self.canvas.coords(key, new_point_x, new_point_y, x, y)
+
+                        self.coordinates[obj_number] = (x, y)
+
+                        if self.show_weight.get():
+                            x_medium,y_medium = self.medium_coordinates(new_point_x,new_point_y,x,y)
```
Comment (MorozVladislav) [inline app.py]: PEP8

File: app.py [hunk 20]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+                            self.canvas.coords(self.canvas_obj[self.OBJ_TYPE.WEIGHT][values['weight_id']]['text_id'], x_medium, y_medium)
+                            r = int(self.r / 2) * len(str(values['weight']))
+                            self.canvas.coords(values['weight_id'],x_medium-r, y_medium-r,x_medium+r, y_medium+r)
```
Comment (MorozVladislav) [inline app.py]: PEP8

File: app.py [hunk 21]
```diff
@@ -162,17 +177,61 @@ def show_weights(self):
...
+
+                break
```
Comment (MorozVladislav) [inline app.py]: This is quite strange. Why do you need brake here?

**Review by MorozVladislav — COMMENTED** [review]
Comment by comment please revise your module

**Review by JuliaVasilkova — APPROVED** [review]

**Pull request merged**